src/catnap/runtime.rs

```diff
@@ -123,40 +124,45 @@ impl MemoryRuntime for PosixRuntime {
         })
     }
 
-    /// Releases a [dmtr_sgarray_t].
+    /// Releases a scatter-gather array.
     fn free_sgarray(&self, sga: dmtr_sgarray_t) -> Result<(), Fail> {
-        assert_eq!(sga.sga_numsegs, 1);
-        for i in 0..sga.sga_numsegs as usize {
-            let seg: &dmtr_sgaseg_t = &sga.sga_segs[i];
-            let allocation: Box<[u8]> = unsafe {
-                Box::from_raw(slice::from_raw_parts_mut(
-                    seg.sgaseg_buf as *mut _,
-                    seg.sgaseg_len as usize,
-                ))
-            };
-            drop(allocation);
+        // Bad scatter-gather.
+        // TODO: Drop this check once we support scatter-gather arrays with multiple segments.
+        if sga.sga_numsegs != 1 {
+            return Err(Fail::new(
+                libc::EINVAL,
+                "scatter-gather array with invalid size",
+            ));
         }
 
+        // Release heap-managed buffer.
+        let sgaseg: dmtr_sgaseg_t = sga.sga_segs[0];
+        let (data_ptr, length): (*mut u8, usize) =
+            (sgaseg.sgaseg_buf as *mut u8, sgaseg.sgaseg_len as usize);
+
+        // Convert back raw slice to a heap buffer and drop allocation.
+        DataBuffer::from_raw_parts(data_ptr, length)?;
+
         Ok(())
     }
 
-    /// Clones a [dmtr_sgarray_t] into a memory buffer.
-    fn clone_sgarray(&self, sga: &dmtr_sgarray_t) -> Result<Bytes, Fail> {
-        let mut len: u32 = 0;
-        for i in 0..sga.sga_numsegs as usize {
-            len += sga.sga_segs[i].sgaseg_len;
+    /// Clones a scatter-gather array.
+    fn clone_sgarray(&self, sga: &dmtr_sgarray_t) -> Result<DataBuffer, Fail> {
+        // Bad scatter-gather.
+        // TODO: Drop this check once we support scatter-gather arrays with multiple segments.
+        if sga.sga_numsegs != 1 {
+            return Err(Fail::new(
+                libc::EINVAL,
+                "scatter-gather array with invalid size",
+            ));
         }
-        let mut buf: BytesMut = BytesMut::zeroed(len as usize).unwrap();
-        let mut pos: usize = 0;
-        for i in 0..sga.sga_numsegs as usize {
-            let seg: &dmtr_sgaseg_t = &sga.sga_segs[i];
-            let seg_slice = unsafe {
-                slice::from_raw_parts(seg.sgaseg_buf as *mut u8, seg.sgaseg_len as usize)
-            };
-            buf[pos..(pos + seg_slice.len())].copy_from_slice(seg_slice);
-            pos += seg_slice.len();
-        }
-        Ok(buf.freeze())
+
+        let sgaseg: dmtr_sgaseg_t = sga.sga_segs[0];
+        let (ptr, len): (*mut c_void, usize) = (sgaseg.sgaseg_buf, sgaseg.sgaseg_len as usize);
+
+        // Clone heap-managed buffer.
+        let seg_slice: &[u8] = unsafe { slice::from_raw_parts(ptr as *const u8, len) };
+        Ok(DataBuffer::from_slice(seg_slice))
```

Comment:
Instead of constructing a slice here, and calling DataBuffer::from_slice(), could you just call DataBuffer::from_raw_parts() directly?  Or would that do something different?

Reply:
There is no semantic difference between what you suggested from what I did. It's all about the style that I like to follow - avoid one-liners. Anyways I merged the code and you asked.